File: functions/getVideo.js
// The Refactored Serverless Function: Simple, Robust, Focused.
const ytdl = require('ytdl-core');

exports.handler = async (event) => {
    const { videoId } = event.queryStringParameters;

    if (!videoId) {
        return { statusCode: 400, body: JSON.stringify({ error: 'Missing videoId.' }) };
    }

    const youtubeUrl = `https://www.youtube.com/watch?v=${videoId}`;
    console.log(`Requesting video+audio stream for: ${videoId}`);

    try {
        const info = await ytdl.getInfo(youtubeUrl);
        
        // We ONLY request a format with both video and audio.
        const format = ytdl.chooseFormat(info.formats, {
            quality: 'highest',
            filter: 'videoandaudio' 
        });
        
        if (!format) {
            throw new Error(`No suitable video+audio format found.`);
        }

        return {
            statusCode: 200,
            body: JSON.stringify({ streamUrl: format.url }),
        };

    } catch (error) {
        console.error("ytdl-core FAILURE:", error);
        return {
            statusCode: 500,
            body: JSON.stringify({ 
                error: 'Could not fetch the video stream.',
                details: error.message
            }),
        };
    }
};

File: functions/getVideo.py
import json
import yt_dlp

def handler(event, context):
    try:
        query_params = event.get('queryStringParameters', {})
        video_id = query_params.get('videoId')

        if not video_id:
            return { 'statusCode': 400, 'body': json.dumps({ 'error': 'Missing videoId.' }) }

        youtube_url = f"https://www.youtube.com/watch?v={video_id}"

        ydl_opts = {
            'format': 'bestvideo[ext=mp4]+bestaudio[ext=m4a]/best[ext=mp4]/best',
            'quiet': True,
        }

        with yt_dlp.YoutubeDL(ydl_opts) as ydl:
            info = ydl.extract_info(youtube_url, download=False)
            stream_url = info.get('url')

            if not stream_url:
                raise Exception('No stream URL found by yt-dlp.')

            return {
                'statusCode': 200,
                'body': json.dumps({ 'streamUrl': stream_url })
            }

    except Exception as e:
        # This will now log the true Python error to your Netlify log
        print(f"CRITICAL PYTHON FAILURE: {e}")
        return {
            'statusCode': 500,
            'body': json.dumps({ 'error': 'The Python back-end failed.', 'details': str(e) })
        }
